test/framework/src/main/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java

```diff
@@ -224,21 +225,41 @@ public static boolean isLinearizable(SequentialSpec spec, History history, Funct
      * @param spec the sequential specification of the datatype
      * @param history the history of events to check for linearizability
      * @param missingResponseGenerator used to complete the history with missing responses
-     * @param terminateEarly a condition upon which to terminate early
      * @return true iff the history is linearizable w.r.t. the given spec
      */
-    public static boolean isLinearizable(
-        SequentialSpec spec,
-        History history,
-        Function<Object, Object> missingResponseGenerator,
-        BooleanSupplier terminateEarly
-    ) {
-        history = history.clone(); // clone history before completing it
-        history.complete(missingResponseGenerator); // complete history
-        final Collection<List<Event>> partitions = spec.partition(history.copyEvents());
-        return partitions.stream().allMatch(h -> isLinearizable(spec, h, terminateEarly));
+    public static boolean isLinearizable(SequentialSpec spec, History history, Function<Object, Object> missingResponseGenerator)
+        throws LinearizableCheckAborted {
+        final ScheduledThreadPoolExecutor scheduler = Scheduler.initScheduler(Settings.EMPTY, "test-scheduler");
+        final AtomicBoolean abort = new AtomicBoolean();
+        try {
+            history = history.clone(); // clone history before completing it
+            history.complete(missingResponseGenerator); // complete history
+            final Collection<List<Event>> partitions = spec.partition(history.copyEvents());
+            // Large histories can be problematic and have the linearizability checker run OOM
+            // Bound the time how long the checker can run on such histories (Values empirically determined)
+            if (history.size() > 300 || partitions.stream().anyMatch(p -> p.size() > 25)) {
+                logger.warn("Detected large history or partition for linearizable check. Limiting execution time");
+                scheduler.schedule(() -> abort.set(true), 10, TimeUnit.SECONDS);
+            }
+            var allLinearizable = partitions.stream().allMatch(h -> isLinearizable(spec, h, abort::get));
+            if (abort.get()) {
+                throw new LinearizableCheckAborted();
+            }
+            return allLinearizable;
+        } catch (OutOfMemoryError e) {
```

Comment:
Unfortunately an OOME might mean that some data structures are in an undefined state so I don't think it's ok to catch it and try to continue with the rest of the tests. Possibly one could argue that it's ok if the only data structures we could be mutating are discarded at this point, but I'm not sure that's true and even if it is true today it seems like it would be a fairly fragile argument given all the other things that might happen concurrently in these tests.

Reply:
Will remove